Remove getTitle from component tile api

scytacki opened the pull request:
The initial goal of this was to remove the need for the getTitle call in the component TileAPI. The scope expanded as I reviewed how the titles were used. There still could be more clean up here around how unique titles are assigned to tiles, but that will have to wait for another day.

User visible changes:
- **more uniform handling of title from dataSets**: The table and plot (graph) provide the the dataSet name for their title if their title has not been explicitly set. This logic was only used in a few places before. It should now apply everywhere.  However in most cases these tiles have their title set when they are first created so this logic only really applies when the tiles have been manually created by JSON.
- **DataFlow tile bug fixed** DataFlow hides the link button when the title is being edited. Previously it would not re-show the link button if title was set to blank which effectively canceled the tile edit.
- **Removed custom DataFlow rename log message** when the title is changed CLUE logs a generic RENAME message, there is no need for DataFlow to log its own message.

Code Changes:
- new `TileModel.computedTitle` view, this is used to get the title in the TileModel or if not set then the title provided by the content via `.contentTitle`.
- `tileApiInterface?.getTileApi(id)?.getTitle?.()` -> `content?.getTile(id)?.computedTitle`
- getUniqueTitle was simplified because it no longer had to go up to the component api to get the correct title for a tile.
- `useContext(TileModelContext)` is used to replace passing the tileModel to some components
- the size and scale properties passed to the `EditableTileTitle` were not actually needed. This component was using them to compute a css `left` property, but this property was being ignored because the element was position via flex box.
- `EditableTileTile` now works with the tile model directly and sets the title on the model. It still provides `onEndEdit` so parents can do additional logic after editing is done.
- Tile Content components no longer have to provide `getTitle` to the tile api.
- ImageTile was switched to use `BasicEditableTileTitle` it seemed to be doing the same thing. There were some custom width calculations but these were ignored see the css `left` note above.  I'd guess there might have been a desire for the image tile to center the title over the image instead of the whole title.  But this hasn't been working for a while. If we need that feature we should be able to come up with a css based approach.
- removed unused properties from `EditableTableTitle`. I could see how these used to match up with how DataFlow and Geometry work with a link button next to the title. But that link button has been removed.
- the table's `useTileApi` was updated to have `const tileApi = useMemo<ITileApi>(() => ({` instead of `const tileApi: ITileApi = useMemo(() => ({`. This then enforces the type of the argument to useMemo and also makes this functions in this instance of `ITileAPI` show up as a "references" in VSCode when searching.
- renamed `tileModelHooks` to `tileContentAPIActions` because now there is also `tileContentAPIViews`. These are basically typescript helpers. They wrap the objects passed to MST's `actions` and `views` so these objects have the right type, and so these actions and views will show up when searching for references of this API.
- fixed a test that now requires the TileModelContext

kswenson commented on src/models/tiles/tile-model.ts:
```diff
@@ -151,6 +149,30 @@ export const TileModel = types
+    // internally we store the title in the `internalTitle` property
+    // this way we can provide view that computes the title
+    // when serialized we save it out simply as `title`
+    const {title, ...others} = clone(snapshot);
```
I don't think you need to `clone` here. The destructuring will create a new object automatically. The same is true below in the `postProcessor`.

scytacki replied:
True. Good catch. I was being paranoid about not modifying the snapshot in place, but you're right the clone isn't needed.

kswenson commented on src/models/history/tree.ts:
```diff
@@ -58,7 +58,9 @@ export const Tree = types.model("Tree", {
             // This tile has already been added to our list
           } else {
-            const tile = resolveIdentifier(TileModel, document, tileId);
+            // HACK: the snapshot processor returns a type which isn't compatible with
+            // resolveIdentifier
+            const tile = resolveIdentifier(TileModel as IAnyModelType, document, tileId);
```
Given that the main advantage of the `snapshotProcessor` is supposed to be better typing, but the net effect seems to be that it gets the types wrong so that we have to cast in places where we didn't used to have to cast, is it really an improvement? 🤷 And that's without getting into the bug that you found in which the underlying type is inexplicably modified.

scytacki replied:
This is the only place the types are messed up that I remember. Using the other approach caused many more typing problems. For example `TileModel.create({title: "me"})` did not work. 

That bug is annoying but in this case we aren't using the underlying internal type, so it won't affect us.

scytacki commented:
@tealefristoe I added you a reviewer just because you like PRs that remove code. 😉 
@bgoldowsky I added you because it might be useful for you know more about this title stuff in the future.
@kswenson you commented before 😄, and I reverted the change the commented on.

kswenson reviewed (approved):
👍 Looks good -- I think avoiding the SnapshotProcessor is a good call.

scytacki commented on src/components/tiles/geometry/geometry-content.tsx:
```diff
@@ -56,6 +56,7 @@ import ErrorAlert from "../../utilities/error-alert";
 import { halfPi, normalizeAngle, Point } from "../../../utilities/math-utils";
 import SingleStringDialog from "../../utilities/single-string-dialog";
 import { getClipboardContent, pasteClipboardImage } from "../../../utilities/clipboard-utils";
+import { ToolTitleArea } from "../tile-title-area";
```
Good point. I'll make it part of the next PR. I don't want to update this one right now since it would take another 100 minutes to prove that nothing was broken.

bgoldowsky commented:
You dont' need my approval but this was interesting to read through.  Looks like a good direction to make things a little simpler and more consistent. 👍

scytacki commented:
@tejal-shah I don't know if this really needs QA review, but I added just in case. It did pass the regression tests when I first added `run regression` label. Now that I've added `pending QA review` that replaced results shown in the PR.  Here is the passing test run: https://github.com/concord-consortium/collaborative-learning/actions/runs/6339741689

tealefristoe reviewed (approved):
I didn't check every changed line, but I liked everything I saw 👍

tejal-shah reviewed (approved):
LGTM 👍
